src/senaite/core/catalog/indexer/sample.py

```diff
@@ -10,19 +10,37 @@
 
 @indexer(IAnalysisRequest)
 def assigned_state(instance):
-    """Returns `assigned` or `unassigned` depending on the state of the
-    analyses the analysisrequest contains. Return `unassigned` if the Analysis
-    Request has at least one analysis in `unassigned` state.
-    Otherwise, returns `assigned`
+    """Returns `assigned`, `unassigned` or 'not_applicable' depending on the
+    state of the analyses the analysisrequest contains. Return `unassigned` if
+    the Analysis Request has at least one 'active' analysis in `unassigned`
+    status. Returns 'assigned' if all 'active' analyses of the sample are
+    assigned to a Worksheet. Returns 'not_applicable' if no 'active' analyses
+    for the given sample exist
     """
-    analyses = instance.getAnalyses()
-    if not analyses:
-        return "unassigned"
-    for analysis in analyses:
-        analysis_object = api.get_object(analysis)
-        if not analysis_object.getWorksheet():
+    assigned = False
+    skip_statuses = ["retracted", "rejected", "cancelled"]
+    for analysis in instance.getAnalyses():
```

Comment:
I think we should stop in custom index functions to rely on catalog queries and brain metadata, but rather always operate on full objects. In this specific case this means to rather use `instance.objectValues(spec="Analysis")` than using the accessor, which does another catalog query as far as I remember.
The reason is, that we never know if the sample is indexed before the analyses or any other dependent catalog.

Reply:
Yes, I fully agree. Because of same reasoning, I'd say accesors should never perform catalog queries, except for when UIDs are stored and return of full objects is expected. 

Therefore, I feel would be better to rely on objects contained in the getter itself (https://github.com/senaite/senaite.core/blob/d1c46a22688b869c625bf165daeb87b61ff42e65/src/bika/lims/browser/fields/aranalysesfield.py#L68) rather than here. What do you think?

Reply:
Yes, I thought about this as well. But this would mean to also refactor all code places where we call `getAnalyses()` without arguments and ensure that these functions do not access the metadata columns. It's a pity that it is not the other way around, so that we have a parameter to return brains.

Reply:
Anyhow, this is a story of another PR. So we merge this and fix it in the setter instead, right?

Reply:
yes, I was looking at this right now and indeedm, is a big refactoring. Don't think this is reasonable for this fixes version we have atm. Will simply stick to the objectvalues in the indexer and keep this refactor in mind

Reply:
Problem is that if I don't rely on the getter, but on `objectValues` I have to take into consideration the sample partitions as well (note the getter performs a query with param `getAncestorsUIDs`). And for the retrieval of the partitions from a sample, a query is also required because "ParentAnalysisRequest" field stores the parent instead of the children (partitions). Thus, a backrefs query against the reference catalog is required! :( . OMG, if at least this field was not from a `ReferenceField` type, but from a `UIDReferenceField`, the query would be against `uid_catalog`, that I believe is more reliable.